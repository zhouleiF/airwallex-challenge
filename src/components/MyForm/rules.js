// email 校验规则
export const regEmail = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
