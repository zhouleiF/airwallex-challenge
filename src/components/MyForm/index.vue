<template>
  <div class="form">
    <form @submit.prevent="submit">
      <input
        :class="{error: hasCheckedName === false}"
        v-model="name"
        name="name"
        placeholder="Full name"
        @input="emitCheck('Name', 'input')"
        @blur="emitCheck('Name')"
      />
      <input
        :class="{error: hasCheckedEmail === false}"
        v-model="email"
        name="email"
        placeholder="Email"
        @input="emitCheck('Email', 'input')"
        @blur="emitCheck('Email')"
      />
      <input
        :class="{error: hasCheckedConfirm === false}"
        v-model="confirmEmail"
        name="confirmEmail"
        placeholder="Confirm Email"
        @input="emitCheck('Confirm', 'input')"
        @blur="emitCheck('Confirm')"
      />
      <button type="submit" class="btn" :class="{disabled: isLoading}">{{isLoading ? 'Sending, please wait.' : 'Send'}}</button>
      <p class="error" v-show="error">{{error}}</p>
    </form>
  </div>
</template>
<script>
import { regEmail } from './rules'
import request from '@/service'

export default {
  name: 'MyForm',
  data () {
    return {
      name: null,
      email: null,
      confirmEmail: null,
      hasCheckedName: true,
      hasCheckedEmail: true,
      hasCheckedConfirm: true,
      error: null,
      isLoading: false
    }
  },
  methods: {
    submit () {
      let checked = true
      const arr = ['Name', 'Email', 'Confirm']
      arr.forEach(type => {
        const isChecked = this.emitCheck(type)
        checked = checked && isChecked
      })
      if (checked && (!this.isLoading)) {
        this.isLoading = true
        this.error = null
        request({
          name: this.name,
          email: this.email
        }).then(res => {
          this.$emit('afterSubmit')
        }).catch(error => (this.error = error)).finally(() => (this.isLoading = false))
      }
    },
    emitCheck (type = '', eventType = 'blur') {
      if (type) {
        const checked = this['check' + type]()
        if (eventType === 'blur') this['hasChecked' + type] = checked
        else if (checked === true) this['hasChecked' + type] = true
        return checked
      }
    },
    checkName () {
      const value = this.name
      return !!(value && (value.length >= 3))
    },
    checkEmail () {
      const value = this.email
      if (this.confirmEmail) {
        this.emitCheck('Confirm')
      }
      return !!(value && regEmail.test(value))
    },
    checkConfirm () {
      const value = this.confirmEmail
      return !!(value && (value === this.email))
    },
    clearForm () {
      Object.assign(this.$data, this.$options.data())
    }
  }
}
</script>
<style lang="scss" scoped>
.form {
  input {
    -webkit-user-select:text !important;
    display: block;
    width: 100%;
    font-size: 30px;
    line-height: 1.5;
    padding-top: 50px;
    border: none;
    border-bottom: 1PX solid #000;
    &:-webkit-autofill {
      box-shadow: 0 0 0px 1000px #fff inset !important;
    }
    &::-webkit-input-placeholder {
        color: #bdcedb;
      }
    &.error {
      border-bottom-color: #d63449;
      &:-webkit-autofill {
        -webkit-text-fill-color: red;
      }
    }
  }
  .btn {
    display: block;
    width: 100%;
    line-height: 80px;
    border-radius: 50px;
    background: #56BACF;
    color: #fff;
    margin: 80px auto 10px;
    &.disabled {
      background-color: #ccc;
    }
  }
  .error {
    color: red;
  }
}
</style>
