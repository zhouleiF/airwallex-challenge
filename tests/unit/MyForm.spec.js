import { mount } from '@vue/test-utils'
import MyForm from '@/components/MyForm/index.vue'
import flushPromises from 'flush-promises'
import { errorMessage } from './_mocks_/axios'
jest.mock('axios')
const afterSubmit = jest.fn(() => {})

describe('MyForm component test', () => {
  const wrapper = mount(MyForm, {
    listeners: {
      afterSubmit
    }
  })
  const nameInput = wrapper.find('input[name="name"]')
  const emailInput = wrapper.find('input[name="email"]')
  const confirmEmailInput = wrapper.find('input[name="confirmEmail"]')
  const form = wrapper.find('form')
  const errorNode = wrapper.find('.error')
  it('renders MyForm correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('submit directly with no fill, all fields highlight', async () => {
    form.trigger('submit')
    await wrapper.vm.$nextTick()
    expect(nameInput.classes('error')).toBe(true)
    expect(emailInput.classes('error')).toBe(true)
    expect(confirmEmailInput.classes('error')).toBe(true)
  })
  it('full name needs to be at least 3 characters long', async () => {
    nameInput.setValue('zl')
    await wrapper.vm.$nextTick()
    expect(nameInput.classes('error')).toBe(true)
    nameInput.setValue('zhoulei')
    await wrapper.vm.$nextTick()
    expect(nameInput.classes('error')).toBe(false)
  })
  it('email needs to be in validation email format', async () => {
    emailInput.setValue('zhoulei@')
    await wrapper.vm.$nextTick()
    expect(emailInput.classes('error')).toBe(true)
    emailInput.setValue('zhoulei@qq.com')
    await wrapper.vm.$nextTick()
    expect(emailInput.classes('error')).toBe(false)
  })
  it('confirm email needs to match email.', async () => {
    confirmEmailInput.setValue('zl')
    await wrapper.vm.$nextTick()
    expect(confirmEmailInput.classes('error')).toBe(true)
    confirmEmailInput.setValue('zhoulei@qq.com')
    await wrapper.vm.$nextTick()
    expect(confirmEmailInput.classes('error')).toBe(false)
  })
  it('when all fields validate properly, submit successful, show the loading and emit afterSubmit event', async () => {
    form.trigger('submit')
    expect(wrapper.vm.isLoading).toBe(true)
    await flushPromises()
    expect(wrapper.vm.isLoading).toBe(false)
    expect(afterSubmit.mock.calls.length).toBe(1)
  })
  it('app should display error message from the server, case: email has used', async () => {
    emailInput.setValue('usedemail@airwallex.com')
    confirmEmailInput.setValue('usedemail@airwallex.com')
    form.trigger('submit')
    await flushPromises()
    expect(errorNode.text()).toBe(errorMessage)
  })
})
