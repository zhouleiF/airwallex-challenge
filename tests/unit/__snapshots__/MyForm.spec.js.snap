// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyForm component test renders MyForm correctly 1`] = `
<div
  class="form"
>
  <form>
    <input
      class=""
      name="name"
      placeholder="Full name"
    />
     
    <input
      class=""
      name="email"
      placeholder="Email"
    />
     
    <input
      class=""
      name="confirmEmail"
      placeholder="Confirm Email"
    />
     
    <button
      class="btn"
      type="submit"
    >
      Send
    </button>
     
    <p
      class="error"
      style="display: none;"
    >
      
    </p>
  </form>
</div>
`;
